collector/lib/DriverCandidates.cpp

```diff
@@ -0,0 +1,194 @@
+  std::string name = driverFullName(shortName, useEbpf);
+
+  return DriverCandidate(name, useEbpf, shortName);
+}
+
+// The kvm driver for minikube uses a custom kernel built from
+// mostly vanilla kernel headers and its own configuration defined
+// in their repo. However, when using the docker driver, minikube
+// runs directly on the host, so we add the kvm kernel as a candidate
+// in order to give the chance for collector to use the host driver.
+std::optional<DriverCandidate> getMinikubeCandidate(HostInfo& host, bool useEbpf) {
+  auto minikube_version = host.GetMinikubeVersion();
+
+  if (minikube_version.empty()) {
+    return {};
+  }
+
+  auto kernel = host.GetKernelVersion();
+
+  std::string shortName = kernel.ShortRelease() + "-minikube-" + minikube_version;
+  std::string name = driverFullName(shortName, useEbpf);
+  return DriverCandidate(name, useEbpf, shortName);
+}
+
+// Normalizes this host's release string into something collector can use
+// to download appropriate kernel objects from the webserver. If the release
+// string does not require normalization, it is simply returned.
+std::string normalizeReleaseString(HostInfo& host) {
+  auto kernel = host.GetKernelVersion();
+  if (host.IsCOS()) {
+    std::string release = kernel.release;
+    // remove the + from the end of the kernel version
+    if (release[release.size() - 1] == '+') {
+      release.erase(release.size() - 1);
+    }
+    return release + "-" + host.GetBuildID() + "-" + host.GetOSID();
+  }
+
+  if (host.IsDockerDesktop()) {
+    auto smp = kernel.version.find("SMP ");
+    if (smp == std::string::npos) {
+      CLOG(FATAL) << "Unable to parse docker desktop kernel version: "
+                  << "'" << kernel.version << "'";
+    }
+
+    std::string time_string = kernel.version.substr(smp + 4);
+    std::tm tm{};
+    // Currently assuming that all docker desktop kernels have UTC timestamps
+    // to simplify parsing for this edge case. std::get_time does not support parsing
+    // timezone information (%Z)
+    if (strptime(kernel.version.substr(smp + 4).c_str(), "%a %b %d %H:%M:%S UTC %Y", &tm) == nullptr) {
+      CLOG(FATAL) << "Failed to parse DockerDesktop kernel timestamp: '" << time_string << "'";
+    }
+    std::stringstream timestamp;
+    timestamp << std::put_time(&tm, "%Y-%m-%d-%H-%M-%S");
+    return kernel.ShortRelease() + "-dockerdesktop-" + timestamp.str();
+  }
+
+  return kernel.release;
+}
+
+DriverCandidate getHostCandidate(HostInfo& host, bool useEbpf) {
+  std::string hostCandidate = normalizeReleaseString(host);
+  std::string hostCandidateFullName = driverFullName(hostCandidate, useEbpf);
+
+  return DriverCandidate(hostCandidateFullName, useEbpf, hostCandidate);
+}
+
+DriverCandidate getUserDriverCandidate(const char* full_name, bool useEbpf) {
+  std::string_view path(full_name);
+
+  // if the name starts with '/' we assume a fully qualified path was provided
+  if (path.rfind("/", 0) == 0) {
```

Comment:
Would it be too much work to use `std::filesystem::path` from c++17 here? There are few nice methods, like `is_absolute`,  `filename`,  `parent_path`.

Reply:
No, I think it makes sense to explore us using `std::filesystem`, I'll give it a try.

Reply:
Good news and bad news.

Good news is the code is much more readable and simpler using `std::filesystem`.

Bad news, GCC-8 had `std::filesystem` defined in a separate `stdc++fs` library and it took me a while to get this to link properly. GCC-9 had this merged into the `stdc++` library, so I think it's fine for us to keep this, since we will get GCC-11 once we move to Centos Stream/UBI 9.

Reply:
The result looks great!